src/walk-manifest.js

```diff
@@ -49,116 +50,180 @@ var parseManifest = function(content) {
-		parent.content = new Buffer(parent.content.toString().replace(uri, path.relative(path.dirname(parent.file), manifest.file)));
-	}
-
-  manifest.content = syncRequest('GET', manifest.uri).getBody();
-  manifest.parsed  = parseManifest(manifest.content);
-	manifest.parsed.segments = manifest.parsed.segments   || [];
-	manifest.parsed.playlists = manifest.parsed.playlists || [];
-	manifest.parsed.mediaGroups = manifest.parsed.mediaGroups || {};
-
-  var playlists = manifest.parsed.playlists.concat(mediaGroupPlaylists(manifest.parsed.mediaGroups));
-	var key = parseKey(basedir, decrypt, resources, manifest, parent);
-
-	// SEGMENTS
-	manifest.parsed.segments.forEach(function(s, i) {
-		if (!s.uri) {
-			return;
-		}
-		// put segments in manifest-name/segment-name.ts
-		s.file = path.join(path.dirname(manifest.file), fsSanitize(path.basename(s.uri)));
-		if (!isAbsolute(s.uri)) {
-			s.uri = joinURI(path.dirname(manifest.uri), s.uri);
-		}
-		if (key) {
-			s.key = key;
-			s.key.iv = s.key.iv || new Uint32Array([0, 0, 0, manifest.parsed.mediaSequence, i]);
-		}
-
-		manifest.content = new Buffer(manifest.content.toString().replace(
-			s.uri,
-			path.relative(path.dirname(manifest.file), s.file)
-		));
-		resources.push(s);
-	});
-
-	// SUB Playlists
-	playlists.forEach(function(p, z) {
-		if (!p.uri) {
-			return;
-		}
-		resources = resources.concat(walkPlaylist(decrypt, basedir, p.uri, manifest, z));
-	});
-
-	return resources;
+var walkPlaylist = function(options, callback) {
+  var {
+    decrypt,
+    basedir,
+    uri,
+    parent = false,
+    manifestIndex = 0,
+    onError = function(err, uri, resources, callback) {
+      // Avoid adding the top level uri to nested errors
+      if (err.message.includes('|')) {
+        return callback(err);
+      } else {
+        var errWithUri = new Error(err.message + '|' + uri);
+        return callback(errWithUri);
+      }
+    },
+    visitedUrls = []
+  } = options;
+
+  var resources = [];
+  var manifest = {};
+  manifest.uri = uri;
+  manifest.file = path.join(basedir, fsSanitize(path.basename(uri)));
+
+  // if we are not the master playlist
+  if (parent) {
+    manifest.file = path.join(
+      path.dirname(parent.file),
+      'manifest' + manifestIndex,
+      fsSanitize(path.basename(manifest.file))
+    );
+    // get the real uri of this playlist
+    if (!isAbsolute(manifest.uri)) {
+      manifest.uri = joinURI(path.dirname(parent.uri), manifest.uri);
+    }
+    // replace original uri in file with new file path
+    parent.content = new Buffer(parent.content.toString().replace(uri, path.relative(path.dirname(parent.file), manifest.file)));
+  }
+
+  if (visitedUrls.includes(manifest.uri)) {
+    var manifestError = new Error('Trying to visit the same uri again; stuck in a cycle');
+    return onError(manifestError, manifest.uri, resources, callback);
+  }
+
+  request({url: manifest.uri, timeout: 1500}, function(error, response, body) {
```

Comment:
We should use `requestretry` for this since instead of adding another thing, since its used in `write-data`